### DISCUSSION TITLE
Recommended way how to transpile dependencies

### DISCUSSION THREAD

**Discussion by vojtechportes** [vercel/next.js#21312] [Help]
Hi,
I am using `next.js` with `styled-components` and dependencies (node_modules) such as ui libraries that are using `styled-components` inside as an peerDependency and are compiled with tsc and being setup with `babel-plugin-styled-components` and all the recommended settings around it.

Since `next.js` is not transpiling `node_modules`, I am ending up with components that are throwing errors about mismatched class names. The only solution I found is using `next-transpile-modules`. With that all the errors are gone and everything is working as expected. My issue though is that starting next.js app in dev or in production mode is taking forever (around 10 minutes (presumably due to large amount of "components" in that transpiled dependencies and due to large amount of dependencies in general)). Hot reloads are also quite slow. 

So I am wonering. Isn't there a better way how to solve this issue? Maybe by configuring webpack in next.config.js? Which I tried but I couldn't make it work. Any ideas?

My current setup:

next.config.js

```
const withPlugins = require("next-compose-plugins");
const withTM = require("next-transpile-modules")([
  "module1",
  "module2"
]);

module.exports = withPlugins([withTM], {
  webpack: (config) => {
    config.module.rules.push({
      test: /\\.+(js|ts|tsx)$/,
      use: {
        loader: "babel-loader",
        options: {
          presets: ["next/babel"],
          plugins: [
            [
              "styled-components",
              {
                ssr: true,
                displayName: true,
                preprocess: false,
                minify: true,
                transpileTemplateLiterals: true,
                pure: true,
              },
            ],
          ],
        },
      },
    });
  }
);
```

_document.tsx

```
import Document, {
  DocumentContext,
  Head,
  Html,
  Main,
  NextScript,
} from "next/document";
import * as React from "react";
import { ServerStyleSheet } from "styled-components";

export default class RendererDocument extends Document<
  DocumentContext
> {
  static async getInitialProps(context: DocumentContext) {
    const sheet = new ServerStyleSheet();
    const originalRenderPage = context.renderPage;

    try {
      context.renderPage = () =>
        originalRenderPage({
          enhanceApp: App => props => sheet.collectStyles(<App {...props} />),
        });

      const initialProps = await Document.getInitialProps(context);

      return {
        ...initialProps,
        styles: (
          <>
            {initialProps.styles}
            {sheet.getStyleElement()}
          </>
        ),
      };
    } finally {
      sheet.seal();
    }
  }

  render() {
    return (
      <Html>
        <Head>
          <meta charSet="utf-8" />
        </Head>
        <body>
          <Main />
          <NextScript />
        </body>
      </Html>
    );
  }
}

```

**Comment by palakurthivishal** [comment]
Hey, I have come across exactly the same case. Have you found any better solution?

It'd be of great help to me.
